lib/axon/metrics.ex

```diff
@@ -365,6 +365,44 @@ defmodule Axon.Metrics do
     |> Nx.mean()
   end
 
+  @doc ~S"""
+  Computes the top-k categorical accuracy.
+
+  ## Options
+
+    * `k` - The k in "top-k". Defaults to 5.
+
+  ## Argument Shapes
+
+    * `y_true` - $\(d_0, d_1, ..., d_n\)$
+    * `y_pred` - $\(d_0, d_1, ..., d_n\)$
+
+  ## Examples
+
+      iex> Axon.Metrics.top_k_categorical_accuracy(Nx.tensor([0, 1, 0, 0, 0]), Nx.tensor([0.1, 0.4, 0.7, 0.7, 0.1]), k: 2)
+      #Nx.Tensor<
+        f32
+        0.0
+      >
+
+      iex> Axon.Metrics.top_k_categorical_accuracy(Nx.tensor([[0, 1, 0], [1, 0, 0]]), Nx.tensor([[0.1, 0.4, 0.7], [0.1, 0.4, 0.7]]), k: 2)
+      #Nx.Tensor<
+        f32
+        0.5
+      >
+
+  """
+  defn top_k_categorical_accuracy(y_true, y_pred, opts \\ []) do
+    opts = keyword!(opts, k: 5)
```

Comment:
Maybe we can add a `:sparse` case as well? If `sparse: true` you don't need to compute the the argmax of `y_true` and instead just pass the `y_true` direct

Reply:
Oh good point! I'll add it.

Reply:
Okay maybe this is a silly question but how would you go about handling that conditional? I'm struggling to get past `boolean predicate passed to cond` in defn. Would it be reasonable to make the function a normal `def` function that delegates to `defnp` after assigning the index based on that conditional?

Reply:
Perhaps taking a look into Nx.LinAlg.norm or Nx.LinAlg.determinant might provide insight on this conditional implementation :)

Reply:
Interesting! So `norm` is defn that delegates to defp and `determinant` is def that delegates to defnp. I think I'm more confused than where I started now lol.

Reply:
The issue is that you cannot execute regular code inside a `defn`, so you need to do something like this:

```
transform(opts[:flag], fn flag ->
  if flag do
    call_another_defn_for_flag(...)
  else
    call_a_defn_without_flag(...)
  end
end)
```

Reply:
We definitely need to improve this though...

Reply:
Alright, how's that now?